File: app/src/main/java/igrek/songbook/layout/songpreview/view/quickmenu/QuickMenuTranspose.java
package igrek.songbook.layout.songpreview.view.quickmenu;

import android.view.View;
import android.widget.Button;
import android.widget.TextView;

import javax.inject.Inject;

import dagger.Lazy;
import igrek.songbook.R;
import igrek.songbook.dagger.DaggerIoc;
import igrek.songbook.info.UiResourceService;
import igrek.songbook.layout.songpreview.LyricsManager;
import igrek.songbook.layout.songpreview.SongPreviewLayoutController;
import igrek.songbook.layout.songpreview.transpose.ChordsTransposerManager;
import igrek.songbook.layout.songpreview.view.SongPreview;

public class QuickMenuTranspose {
	
	@Inject
	Lazy<LyricsManager> lyricsManager;
	@Inject
	Lazy<ChordsTransposerManager> chordsTransposerManager;
	@Inject
	UiResourceService infoService;
	@Inject
	Lazy<SongPreviewLayoutController> songPreviewController;
	
	private boolean visible = false;
	private View quickMenuView;
	private TextView transposedByLabel;
	
	public QuickMenuTranspose() {
		DaggerIoc.getFactoryComponent().inject(this);
	}
	
	private SongPreview getCanvas() {
		return songPreviewController.get().getSongPreview();
	}
	
	public void setQuickMenuView(View quickMenuView) {
		this.quickMenuView = quickMenuView;
		
		transposedByLabel = quickMenuView.findViewById(R.id.transposedByLabel);
		
		Button btnTransposeM5 = quickMenuView.findViewById(R.id.btnTransposeM5);
		btnTransposeM5.setOnClickListener(v -> chordsTransposerManager.get().onTransposeEvent(-5));
		
		Button btnTransposeM1 = quickMenuView.findViewById(R.id.btnTransposeM1);
		btnTransposeM1.setOnClickListener(v -> chordsTransposerManager.get().onTransposeEvent(-1));
		
		Button btnTranspose0 = quickMenuView.findViewById(R.id.btnTranspose0);
		btnTranspose0.setOnClickListener(v -> chordsTransposerManager.get()
				.onTransposeResetEvent());
		
		Button btnTransposeP1 = quickMenuView.findViewById(R.id.btnTransposeP1);
		btnTransposeP1.setOnClickListener(v -> chordsTransposerManager.get().onTransposeEvent(+1));
		
		Button btnTransposeP5 = quickMenuView.findViewById(R.id.btnTransposeP5);
		btnTransposeP5.setOnClickListener(v -> chordsTransposerManager.get().onTransposeEvent(+5));
		
		updateTranspositionText();
	}
	
	private void updateTranspositionText() {
		String semitonesDisplayName = chordsTransposerManager
				.get()
				.getTransposedByDisplayName();
		String transposedByText = infoService.resString(R.string.transposed_by, semitonesDisplayName);
		transposedByLabel.setText(transposedByText);
	}
	
	public void draw() {
		if (visible) {
			//dimmed background
			float w = getCanvas().getW();
			float h = getCanvas().getH();
			getCanvas().setColor(0x000000, 110);
			getCanvas().fillRect(0, 0, w, h);
		}
	}
	
	public boolean isVisible() {
		return visible;
	}
	
	public void setVisible(boolean visible) {
		this.visible = visible;
		
		if (visible) {
			quickMenuView.setVisibility(View.VISIBLE);
			updateTranspositionText();
		} else {
			quickMenuView.setVisibility(View.GONE);
		}
		
		getCanvas().repaint();
	}
	
	public void onShowQuickMenuEvent(boolean visible) {
		setVisible(visible);
	}
	
	public void onTransposedEvent() {
		if (visible) {
			updateTranspositionText();
		}
	}
	
}

File: app/src/main/java/igrek/songbook/songpreview/quickmenu/QuickMenuTranspose.kt
package igrek.songbook.songpreview.quickmenu

import android.annotation.SuppressLint
import android.view.View
import android.widget.Button
import android.widget.TextView
import igrek.songbook.R
import igrek.songbook.chords.loader.LyricsLoader
import igrek.songbook.chords.model.convertToSharp
import igrek.songbook.chords.syntax.ChordNames
import igrek.songbook.chords.syntax.MajorKey
import igrek.songbook.info.UiResourceService
import igrek.songbook.inject.LazyExtractor
import igrek.songbook.inject.LazyInject
import igrek.songbook.inject.appFactory
import igrek.songbook.settings.preferences.SettingsState

// Singleton
class QuickMenuTranspose(
    lyricsLoader: LazyInject<LyricsLoader> = appFactory.lyricsLoader,
    uiResourceService: LazyInject<UiResourceService> = appFactory.uiResourceService,
    settingsState: LazyInject<SettingsState> = appFactory.settingsState,
) {
    private val lyricsLoader by LazyExtractor(lyricsLoader)
    private val uiResourceService by LazyExtractor(uiResourceService)
    private val preferencesState by LazyExtractor(settingsState)

    var isVisible = false
        set(visible) {
            field = visible
            quickMenuView?.let { quickMenuView ->
                if (visible) {
                    quickMenuView.visibility = View.VISIBLE
                    updateTranspositionText()
                } else {
                    quickMenuView.visibility = View.GONE
                }
            }
        }
    private var quickMenuView: View? = null
    private var transposedByLabel: TextView? = null

    /**
     * @return is feature active - has impact on song preview (panel may be hidden)
     */
    val isFeatureActive: Boolean
        get() = lyricsLoader.isTransposed

    fun setQuickMenuView(quickMenuView: View) {
        this.quickMenuView = quickMenuView

        transposedByLabel = quickMenuView.findViewById(R.id.transposedByLabel)

        val transposeM5Button = quickMenuView.findViewById<Button>(R.id.transposeM5Button)
        transposeM5Button.setOnClickListener {
            lyricsLoader.onTransposeEvent(-5)
        }

        val transposeM1Button = quickMenuView.findViewById<Button>(R.id.transposeM1Button)
        transposeM1Button.setOnClickListener {
            lyricsLoader.onTransposeEvent(-1)
        }

        val transpose0Button = quickMenuView.findViewById<Button>(R.id.transpose0Button)
        transpose0Button.setOnClickListener {
            lyricsLoader.onTransposeResetEvent()
        }

        val transposeP1Button = quickMenuView.findViewById<Button>(R.id.transposeP1Button)
        transposeP1Button.setOnClickListener {
            lyricsLoader.onTransposeEvent(+1)
        }

        val transposeP5Button = quickMenuView.findViewById<Button>(R.id.transposeP5Button)
        transposeP5Button.setOnClickListener {
            lyricsLoader.onTransposeEvent(+5)
        }

        updateTranspositionText()
    }

    @SuppressLint("SetTextI18n")
    private fun updateTranspositionText() {
        val semitonesDisplayName = lyricsLoader.transposedByDisplayName
        val transposedByText =
            uiResourceService.resString(R.string.transposed_by_semitones, semitonesDisplayName)

        val detectedKeyStr = buildSongKeyName(lyricsLoader.songKey)
        val detectedKeyText =
            uiResourceService.resString(R.string.song_detected_key, detectedKeyStr)

        transposedByLabel?.text = "$transposedByText\n$detectedKeyText"
    }

    private fun buildSongKeyName(key: MajorKey): String {
        val notation = preferencesState.chordsNotation
        val forceSharps = preferencesState.forceSharpNotes
        val baseMajorNote = when (forceSharps) {
            true -> convertToSharp(key.baseMajorNote, notation)
            false -> key.baseMajorNote
        }
        val baseMinorNote = when (forceSharps) {
            true -> convertToSharp(key.baseMinorNote, notation)
            false -> key.baseMinorNote
        }
        val majorNoteName = ChordNames.formatNoteName(notation, baseMajorNote, false)
        val minorNoteName = ChordNames.formatNoteName(notation, baseMinorNote, true)
        return "$majorNoteName / $minorNoteName"
    }

    fun onTransposedEvent() {
        if (isVisible) {
            updateTranspositionText()
        }
    }

}
